refactor(codegen): add `CodeBuffer` to fix soundness hole

DonIsaac opened the pull request:
# What This PR Does

Adds `CodeBuffer`, a simple wrapper over a `Vec<u8>` with a protective and reduced API for upholding UTF-8 validity guarantees. Closes #6147.

Note that this struct is actually quite small. Most of the added lines are doc comments.

DonIsaac commented:
* **#6148** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6148?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+///
+/// # Examples
+/// ```
+/// # use oxc_codegen::CodeBuffer;
````
```suggestion
/// use oxc_codegen::CodeBuffer;
```

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+    /// ## Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
````
```suggestion
    /// use oxc_codegen::CodeBuffer;
```

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
```diff
@@ -0,0 +1,422 @@
+    #[inline]
+    #[must_use = "Peeking is pointless if the peeked char isn't used"]
+    pub fn peek_nth(&self, n: usize) -> Option<char> {
+        // SAFETY: `buf` is a valid UTF-8 string because of invariants upheld by
+        // CodeBuffer
```
```suggestion
        // SAFETY: `buf` is a valid UTF-8 string because of invariants upheld by CodeBuffer
```
It looks like better to keep one line

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+    /// # Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
````
```suggestion
    /// use oxc_codegen::CodeBuffer;
```

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+    /// # Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
````
```suggestion
    /// use oxc_codegen::CodeBuffer;
```

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+    /// ## Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
````
```suggestion
    /// use oxc_codegen::CodeBuffer;
```

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+    /// # Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
````
```suggestion
    /// use oxc_codegen::CodeBuffer;
```

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+    /// # Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
````
```suggestion
    /// use oxc_codegen::CodeBuffer;
```

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+    /// # Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
````
```suggestion
    /// use oxc_codegen::CodeBuffer;
```

Dunqing commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,422 @@
+    /// # Examples
+    ///
+    /// ```
+    /// # use oxc_codegen::CodeBuffer;
````
```suggestion
    /// use oxc_codegen::CodeBuffer;
```

Dunqing reviewed (dismissed):
Overall looks good

Boshen commented:
Is this just `String`? @overlookmotel said it could just be a string but I didn't spend the time converting it.

overlookmotel commented on crates/oxc_codegen/src/code_buffer.rs:
```diff
@@ -0,0 +1,421 @@
+        &self.buf
+    }
+}
+impl From<CodeBuffer> for String {
+    fn from(mut printer: CodeBuffer) -> Self {
+        printer.take_source_text()
+    }
+}
```
I'm not sure if this should delegate to `take_source_text`. Unlike in `take_source_text`, here you're consuming the `CodeBuffer`, so no need to `take` its contents.

If compiler inlines `take_source_text`, then it will *probably* see that it can elide the `take` operation. But no guarantee that it will inline, so I think better to play it safe and duplicate the code. And in release mode, this function is a no-op (just a type system cast) so all the code will disappear anyway.

This method should be marked `#[inline]` because it's a no-op.

```suggestion
impl From<CodeBuffer> for String {
    #[inline]
    fn from(printer: CodeBuffer) -> Self {
        #[cfg(debug_assertions)]
        {
            String::from_utf8(printer.buf).unwrap()
        }
        #[cfg(not(debug_assertions))]
        {
            // SAFETY: `buf` is valid UTF-8 because of invariants upheld by `CodeBuffer`
            unsafe { String::from_utf8_unchecked(printer.buf) }
        }
    }
}
```

overlookmotel commented on crates/oxc_codegen/src/lib.rs:
```diff
@@ -569,13 +570,13 @@ impl<'a> Codegen<'a> {
     fn add_source_mapping(&mut self, position: u32) {
         if let Some(sourcemap_builder) = self.sourcemap_builder.as_mut() {
-            sourcemap_builder.add_source_mapping(&self.code, position, None);
+            sourcemap_builder.add_source_mapping(self.code.as_ref(), position, None);
```
Maybe add an `as_bytes` method to do the same? `code.as_ref()` is a slightly unintuitive/mysterious (what kind of ref?)

overlookmotel replied:
I've made this change in an additional commit.

overlookmotel reviewed (commented):
I really like this. As you say, we were providing a safe API which allowed doing unsafe things. But, beyond that, the `str::from_utf8_unchecked` calls in codegen gave me the willies, as without reviewing every line of the entire crate, there was no way to know if it was sound or not. Any code anywhere could sneak in a dodgy byte and violate the safety constraints.

`CodeBuffer` encapsulates the unsafe code and provides a clear contract for what's allowed and what's not. Nice!

Most of my comments below are just nits about naming. I think it's useful to be clear about the difference between bytes and `char`s.

overlookmotel commented:
> Is this just `String`? @overlookmotel said it could just be a string but I didn't spend the time converting it.

Yes, in one sense it's a re-implementation of `String`. This could be a wrapper around a `String` with some added unsafe methods to push bytes to it. But I think the implementation in this PR wrapping a `Vec<u8>` is fine too.

overlookmotel commented on crates/oxc_codegen/src/code_buffer.rs:
```diff
@@ -0,0 +1,421 @@
+    /// [`take_source_text`]: CodeBuffer::take_source_text
+    /// [`print_unchecked`]: CodeBuffer::print_unchecked
+    #[inline]
+    pub unsafe fn print_byte(&mut self, ch: u8) {
```
I think this should be called `print_byte_unchecked` to reflect that it's an unsafe method.

overlookmotel commented:
If you want to remove some of the unsafe calls, you could also copy this: https://doc.rust-lang.org/std/ascii/enum.Char.html

Then to push a space, you'd do:

```rs
self.code.print_ascii(AsciiByte::Space);
```

`AsciiByte` can *only* be an ASCII byte, so no need for an assertion that it is. i.e. Implementation of `print_ascii` would be just `self.buf.push(ascii_byte)` (same as unsafe method `print_byte`), but it can be a safe method.

Also:

```rs
pub(crate) fn print_asciis<I>(&mut self, bytes: I)
    where I: IntoIterator<Item = AsciiByte>,
{
    self.buf.extend(bytes.map(|b| b as u8));
}
```

```rs
// Previously
unsafe { self.code.print_unchecked(std::iter::repeat(b'\t').take(self.indent as usize)); }
// Now
self.code_print_asciis(std::iter::repeat(AsciiByte::Tab).take(self.indent as usize));
```

If you did want to do that, I'd suggest doing it in a follow-up PR.

overlookmotel commented:
One last thing... I was looking at the assembly for the various print methods, and noticed something odd:

`print_char(' ')` and `print_str(" ")` seem to produce tighter assembly than `print_ascii_byte(b' ')`.

https://godbolt.org/z/Kv8sEoEed (ignore all the `alloc::raw_vec::*` stuff at the top - that's all handling `Vec` growth)

`print_ascii_byte` has the "this Vec must grow" branch as the default. That's not ideal because growth strategy is doubling, so e.g. when the `Vec` has just grown from 1024 bytes to 2048 bytes, it won't need to grow again until another 1024 bytes have been pushed. "Needs to grow" is a *very* rare occurrence.

Also, it pushes 3 registers to the stack before it does anything else, and it's quite hefty overall. Not good!

`print_ascii_byte_v2` is my own version which pushes all the "Vec needs to grow" logic into a cold path, and only uses 2 registers.

The question is why does `std::vec::Vec` have this "bad" behavior? This seems odd, as `std` is highly optimized. It's implausible it wouldn't be as it is without a good reason.

My best guess is that `Vec` is optimized for a general use case, and it's not uncommon for a `Vec` to be empty and then have a single item pushed. In which case it only hits the "needs to grow" path, and never hits the "no growth necessary". So my `print_ascii_byte_v2` would penalize this common case.

But for our use case here, we are always pushing a ton of bytes. So it may be that the `print_ascii_byte_v2` would work better for this specific case.

We may also be able to replicate what [`Stack` does](https://github.com/oxc-project/oxc/blob/1cee2070506ec58a8dea86ab9ff3ca3258a98ac3/crates/oxc_transformer/src/helpers/stack/standard.rs) to make all these operations even cheaper. This `push` is 1 operation shorter than `print_ascii_byte_v2`, and uses 1 less register: https://godbolt.org/z/sna4K84G1

This is all micro-optimization stuff, but as pushing bytes to output is such a large component of what `oxc_codegen` does, and its such a hot path, it might be worth giving it a go.

(again, that'd be best done in a follow-up PR)

overlookmotel commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,421 @@
+    /// code.print_ascii(['f', 'o', 'o'].iter().copied());
+    /// assert_eq!(String::from(code), "foo");
+    /// ```
+    pub fn print_ascii<I>(&mut self, chars: I)
````
Don't worry about my "fancy" suggestion. That can be done in a follow-up if we think this function is a perf bottleneck.

overlookmotel replied:
I've changed the var name in an additional commit. This method is never called, so no point optimizing it.

overlookmotel commented:
@DonIsaac Just to check - are any of my comments above blockers for you?

I think this PR is a great improvement, and would very much like to get it merged. Removing potential UB is always a high priority, in my opinion.

If you're busy and would like me to finish it off, please say so.

DonIsaac commented:
> @DonIsaac Just to check - are any of my comments above blockers for you?
> 
> I think this PR is a great improvement, and would very much like to get it merged. Removing potential UB is always a high priority, in my opinion.
> 
> If you're busy and would like me to finish it off, please say so.

None are blockers, I'm just working towards it slowly. I've had a lot going on these past few days 😄

overlookmotel commented on crates/oxc_codegen/src/code_buffer.rs:
````diff
@@ -0,0 +1,421 @@
+    /// ```
+    #[inline]
+    #[must_use = "Peeking is pointless if the peeked char isn't used"]
+    pub fn peek_nth(&self, n: usize) -> Option<char> {
````
Don has made this change.

Boshen commented:
### Merge activity

* **Oct 13, 5:14 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

DonIsaac committed:
refactor(codegen): add `CodeBuffer` to fix soundness hole (#6148)

# What This PR Does

Adds `CodeBuffer`, a simple wrapper over a `Vec<u8>` with a protective and reduced API for upholding UTF-8 validity guarantees. Closes #6147.

Note that this struct is actually quite small. Most of the added lines are doc comments.
